packages/iov-core/src/workers/signingservice.worker.ts

```diff
@@ -0,0 +1,198 @@
+
+  const bnsdConnection = (await signer.addChain(bnsConnector(bnsdUrl))).connection;
+  const bnsdChainId = bnsdConnection.chainId();
+  const ethereumConnection = (await signer.addChain(ethereumConnector(ethereumUrl, undefined))).connection;
+  const ethereumChainId = ethereumConnection.chainId();
+
+  // faucet identity
+  await profile.createIdentity(ed25519Wallet.id, bnsdChainId, HdPaths.simpleAddress(0));
+  // ganache second identity
+  await profile.createIdentity(secp256k1Wallet.id, ethereumChainId, HdPaths.bip44(60, 0, 0, 1));
+
+  const serverCore = new ServerCore(profile, signer);
+
+  async function handleRequest(event: any): Promise<JsonRpcResponse | JsonRpcErrorResponse> {
+    let request: JsonRpcRequest;
+    try {
+      request = parseJsonRpcRequest(event.data);
+    } catch (error) {
+      const requestId = parseJsonRpcId(event.data);
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: requestId,
+        error: {
+          code: jsonRpcCodeInvalidRequest,
+          message: error.toString(),
+        },
+      };
+      return errorResponse;
+    }
+
+    let call: RpcCall;
+    try {
+      call = parseRpcCall(request);
+    } catch (error) {
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: request.id,
+        error: {
+          code: jsonRpcCodeMethodNotFound,
+          message: error.toString(),
+        },
+      };
+      return errorResponse;
+    }
+
+    try {
+      let response: JsonRpcResponse;
+      switch (call.name) {
+        case "getIdentities":
+          response = {
+            jsonrpc: "2.0",
+            id: request.id,
+            result: await serverCore.getIdentities(call.reason, call.chainIds),
+          };
+          break;
+        case "signAndPost":
+          response = {
+            jsonrpc: "2.0",
+            id: request.id,
+            result: await serverCore.signAndPost(call.reason, call.transaction),
+          };
+          break;
+        default:
+          throw new Error("Unsupported RPC call");
+      }
+      return response;
+    } catch (error) {
+      let errorCode: number;
+      if (error instanceof ParamsError) {
+        errorCode = jsonRpcCodeInvalidParams;
+      } else if (error instanceof MethodNotFoundError) {
+        errorCode = jsonRpcCodeMethodNotFound;
+      } else {
+        errorCode = jsonRpcCodeServerErrorDefault;
+      }
+
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: request.id,
+        error: {
+          code: errorCode,
+          message: error.toString(),
+        },
+      };
+      return errorResponse;
+    }
+  }
+
+  onmessage = async event => {
+    // console.log("Received message", JSON.stringify(event));
+
+    // filter out empty {"isTrusted":true} events
+    if (!event.data) {
+      return;
+    }
+
+    const response = await handleRequest(event);
+    postMessage(response);
```

Comment:
Huh? where does this come from?
Ah, this is webworker built-in, right?
Curious how this is matched with the request

Reply:
The globals `postMessage` and `onmessage` are available in the WebWorker API, which is added at compile time via `/// <reference lib="webworker" />` and added by the browser at runtime. This messaging system has no request-response semantics. You can push messages in both directions at any time. JSON-RPC will match those messages using the `id` property in JSON-RPC requests and responses.